### ISSUE TITLE
Rule Proposal: vue/no-unused-methods

### ISSUE THREAD

**Issue by phaibin** [vuejs/eslint-plugin-vue#848]
**Please describe what the rule should do:**

Unused methods should be warned and you can clear them out. 

**What category should the rule belong to?**

- [x] Enforces code style
- [ ] Warns about a potential error
- [ ] Suggests an alternate way of doing something
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule should warn about:**
```vue
<template>
  <h1>Hello</h1>
</template>
<script>
export default {
  data() {
    return {
    }
  },
  mounted() {
    this.init()
  },
  methods: {
    init() {

    },
    unusedMethod() {

    }    
  }
}
</script>
```

This `unusedMethod` method should be set as `Warning` and easily to find out.

**Additional context**

**Comment by mysticatea** [comment]
Thank you for this issue.

However, I'm not sure if this rule works intentionally because methods can be used from outside of the component via `ref`. For example, people may define `focus()` method as a public API.

**Comment by phaibin** [comment]
Thanks for reply. 

I agree the situation you mentioned exists, but I think it's very few. So this rule can be added to `Uncategorized` group and added to eslint config alone. 

At least in developing stage this rule is very useful because you often make a lot of changes for a component and the codes will soon be mess. I think one option is to adding disable tag for a public method or disable this rule for production eslint config.

Anyway, I still looking forward to get this rule worked. 😉

**Comment by ekosynth** [comment]
Any progress on this? Would be extremely useful for me.

**Comment by agm1984** [comment]
I just Googled to find this because I have trouble locating dead code when it is computed props and methods.

I think there should be an ES Lint rule that catches these, and I think if you are using the methods outside the component, you should use `/* eslint-disable-line no-unused-stuff */` along with a code annotation that indicates the method has an implicitly depended on by unknown external code.

**Comment by dten** [comment]
Can we use style to consider methods private to the component an warn if unused?

https://vuejs.org/v2/style-guide/#Private-property-names-essential

**Comment by ota-meshi** [comment]
We have released a rule to report unused properties.

- [vue/no-unused-properties](https://eslint.vuejs.org/rules/no-unused-properties.html)

You can also use the option to report unused methods.

https://eslint.vuejs.org/rules/no-unused-properties.html#options

However, as [as commented](https://github.com/vuejs/eslint-plugin-vue/issues/848#issuecomment-470413303), this rule can't track usage via `$refs`, which might be a false positive.

**Comment by heohex** [comment]
@ota-meshi Hello, is it possible to make [vue/no-unused-properties](https://eslint.vuejs.org/rules/no-unused-properties.html) work with a vue-class-component?
